src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -2,214 +2,151 @@
+                destination[0] = 0x40;   // Literal name
+                if (IntegerEncoder.Encode(index, 6, destination, out int indexLength))
+                {
+                    if (indexLength < destination.Length &&
+                        EncodeAsciiString(value, destination.Slice(indexLength), out int nameLength))
+                    {
+                        bytesWritten = indexLength + nameLength;
+                        return true;
+                    }
+                }
             }
 
-            i += nameLength;
-
-            length = i;
-            return true;
+            bytesWritten = 0;
+            return false;
         }
 
-        public static bool EncodeHeader(string name, string value, Span<byte> buffer, out int length)
+        public static byte[] EncodeIndexedNameToAllocatedArray(int index, ReadOnlySpan<byte> value)
         {
-            int i = 0;
-            length = 0;
-
-            if (buffer.Length == 0)
-            {
-                return false;
-            }
-
-            buffer[i++] = 0;
-
-            if (i == buffer.Length)
-            {
-                return false;
-            }
-
-            if (!EncodeString(name, buffer.Slice(i), out int nameLength, lowercase: true))
-            {
-                return false;
-            }
-
-            i += nameLength;
-
-            if (i >= buffer.Length)
-            {
-                return false;
-            }
-
-            if (!EncodeString(value, buffer.Slice(i), out int valueLength, lowercase: false))
-            {
-                return false;
-            }
-
-            i += valueLength;
-
-            length = i;
-            return true;
+            Span<byte> span = stackalloc byte[256];
+            bool success = EncodeIndexedName(index, value, span, out int length);
+            Debug.Assert(success, "Stack-allocated space was too small to accomodate known name/value.");
+            return span.Slice(0, length).ToArray();
         }
 
-        private static bool EncodeString(string s, Span<byte> buffer, out int length, bool lowercase)
+        public static bool EncodeHeaderNameValue(string name, string value, Span<byte> destination, out int bytesWritten)
         {
-            const int toLowerMask = 0x20;
-
-            int i = 0;
-            length = 0;
-
-            if (buffer.Length == 0)
-            {
-                return false;
-            }
-
-            buffer[0] = 0;
-
-            if (!IntegerEncoder.Encode(s.Length, 7, buffer, out int nameLength))
+            if ((uint)destination.Length > 1 &&
+                EncodeString(name, destination.Slice(1), out int nameLength, toLower: true))
             {
-                return false;
-            }
-
-            i += nameLength;
-
-            // TODO: use huffman encoding
-            for (int j = 0; j < s.Length; j++)
-            {
-                if (i >= buffer.Length)
+                destination[0] = 0;
```

Comment:
It's slightly weird in terms of code readability that we set this byte after we've encoded the string that follows it. Not a big deal, but possibly consider reordering this.